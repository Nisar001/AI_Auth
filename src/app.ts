import 'reflect-metadata';
import express from 'express';
import cors from 'cors';
import dotenv from 'dotenv';
import { AppDataSource } from './config/db';
import { errorHandler, notFoundHandler } from './middlewares/errorHandler';
import { generalLimiter } from './middlewares/rateLimiter';
import authRoutes from './modules/routes/authRoutes';
import logger from './utils/logger';

// Load environment variables
dotenv.config();

const app = express();

// Security and parsing middleware
app.use(cors({
  origin: process.env.CORS_ORIGINS?.split(',') || ['http://localhost:3000'],
  credentials: true,
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'],
  allowedHeaders: ['Content-Type', 'Authorization']
}));

app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));

// Apply rate limiting to all requests
app.use(generalLimiter);

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    environment: process.env.NODE_ENV || 'development'
  });
});

// API routes
app.use('/api/auth', authRoutes);

// Handle 404 errors
app.use(notFoundHandler);

// Global error handler
app.use(errorHandler);

// Initialize database connection
const initializeDatabase = async () => {
  try {
    await AppDataSource.initialize();
    logger.info('Database connected successfully');
  } catch (error) {
    logger.error('Database connection failed:', error);
    process.exit(1);
  }
};

// Graceful shutdown
process.on('SIGTERM', async () => {
  logger.info('SIGTERM received, shutting down gracefully');
  try {
    await AppDataSource.destroy();
    logger.info('Database connection closed');
    process.exit(0);
  } catch (error) {
    logger.error('Error during shutdown:', error);
    process.exit(1);
  }
});

export { app, initializeDatabase };