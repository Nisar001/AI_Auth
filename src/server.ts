import { app, initializeDatabase } from './app';
import logger from './utils/logger';

const PORT = parseInt(process.env.PORT || '3000', 10);

const startServer = async () => {
  try {
    // Initialize database connection
    await initializeDatabase();
    
    // Start the server
    const server = app.listen(PORT, () => {
      logger.info(`🚀 Server running on port ${PORT}`);
      logger.info(`📚 Environment: ${process.env.NODE_ENV || 'development'}`);
      logger.info(`🔗 Health check: http://localhost:${PORT}/health`);
      logger.info(`🔐 Auth API: http://localhost:${PORT}/api/auth`);
    });

    // Handle server errors
    server.on('error', (error: any) => {
      if (error.code === 'EADDRINUSE') {
        logger.error(`Port ${PORT} is already in use`);
      } else {
        logger.error('Server error:', error);
      }
      process.exit(1);
    });

    // Graceful shutdown on SIGINT
    process.on('SIGINT', () => {
      logger.info('SIGINT received, shutting down gracefully');
      server.close(() => {
        logger.info('Server closed');
        process.exit(0);
      });
    });

  } catch (error) {
    logger.error('Failed to start server:', error);
    process.exit(1);
  }
};

// Start the server
startServer();