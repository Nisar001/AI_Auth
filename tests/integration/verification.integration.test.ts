// Correct syntax and scope issues in mocks and test setup
jest.mock('twilio', () => {
  return jest.fn().mockImplementation(() => ({
    messages: { create: jest.fn().mockResolvedValue({ sid: 'mock-sid' }) }
  }));
});

jest.mock('nodemailer', () => {
  return {
    createTransport: jest.fn(() => ({
      sendMail: jest.fn().mockResolvedValue({ messageId: 'mock-message-id' }),
      verify: jest.fn().mockResolvedValue(true)
    }))
  };
});

// Persistent in-memory user store for this test file
let users: any[] = [];
let idCounter = 1;

// Reset in-memory store before each test
beforeEach(() => {
  users = [];
  idCounter = 1;
});

// Replace real DB initialization and destruction with mocks
jest.mock('../../src/config/db', () => {
  return {
    AppDataSource: {
      getRepository: jest.fn(() => {
        return {
          findOne: jest.fn(async (query: any) => {
            if (query.where.email) {
              return users.find(u => u.email === query.where.email) || null;
            }
            if (query.where.phone && query.where.countryCode) {
              return users.find(u => u.phone === query.where.phone && u.countryCode === query.where.countryCode) || null;
            }
            if (query.where.phone) {
              return users.find(u => u.phone === query.where.phone) || null;
            }
            return null;
          }),
          find: jest.fn(async () => users),
          save: jest.fn(async (entity: any) => {
            users.push(entity);
            return entity;
          }),
          count: jest.fn(() => users.length),
        };
      }),
      initialize: jest.fn().mockResolvedValue(true),
      synchronize: jest.fn().mockResolvedValue(true),
    },
  };
});

jest.mock('../../src/services/otpService', () => {
  return {
    OtpService: jest.fn().mockImplementation(() => {
      return {
        generateAndSendOtp: jest.fn().mockResolvedValue({ success: true, otpId: 'mock-otp-id' }),
      };
    }),
  };
});

import request from 'supertest';
import { app } from '../../src/app';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.test' });

// Reset in-memory store before each test
beforeEach(() => {
  users = [];
  idCounter = 1;
});

describe('Resend OTP Integration Tests', () => {
  it('should resend email OTP successfully', async () => {
    // Mock user registration
    await request(app)
      .post('/api/auth/register')
      .send({
        email: 'test@example.com',
        password: 'Test@1234',
        phone: '+1234567890',
        countryCode: '+1',
        fname: 'Test',
        lname: 'User',
        dob: '1990-01-01',
        address: {
          houseNumber: '5E',
          street: 'Fifth St',
          line1: '202 Main St',
          city: 'Testtown',
          state: 'TS',
          country: 'Testland',
          pincode: '987654'
        }
      });

    const res = await request(app)
      .post('/api/auth/resend-otp')
      .send({ identifier: 'test@example.com', type: 'email' });

    expect(res.status).toBe(200);
    expect(res.body.success).toBe(true);
    expect(res.body).toMatchSnapshot({
      data: expect.any(Object),
      message: expect.any(String),
      statusCode: expect.any(Number),
      success: expect.any(Boolean),
    });
  });

  it('should resend phone OTP successfully', async () => {
    const res = await request(app)
      .post('/api/auth/resend-otp')
      .send({ identifier: '+1234567890', type: 'sms' });

    expect(res.status).toBe(200);
    expect(res.body.success).toBe(true);
    expect(res.body).toMatchSnapshot({
      data: expect.any(Object),
      message: expect.any(String),
      statusCode: expect.any(Number),
      success: expect.any(Boolean),
    });
  });

  it('should return error for invalid identifier', async () => {
    const res = await request(app)
      .post('/api/auth/resend-otp')
      .send({ identifier: 'invalid@example.com', type: 'email' });

    expect(res.status).toBe(200);
    expect(res.body.success).toBe(true);
    expect(res.body).toMatchSnapshot({
      message: expect.any(String),
    });
  });

  it('should enforce rate limiting for OTP requests', async () => {
    await request(app)
      .post('/api/auth/resend-otp')
      .send({ identifier: 'test@example.com', type: 'email' });

    const res = await request(app)
      .post('/api/auth/resend-otp')
      .send({ identifier: 'test@example.com', type: 'email' });

    expect(res.status).toBe(429);
    expect(res.body.success).toBe(false);
    expect(res.body).toMatchSnapshot();
  });
});

// Ensure proper cleanup of mocks and open handles
afterEach(() => {
  jest.clearAllMocks();
});
