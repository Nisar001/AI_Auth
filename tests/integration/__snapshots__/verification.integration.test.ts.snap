// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resend OTP Integration Tests should enforce rate limiting for OTP requests 1`] = `
{
  "data": null,
  "message": "Too many OTP requests, please try again in 5 minutes",
  "statusCode": 429,
  "success": false,
}
`;

exports[`Resend OTP Integration Tests should resend email OTP successfully 1`] = `
{
  "data": Any<Object>,
  "message": Any<String>,
  "statusCode": Any<Number>,
  "success": Any<Boolean>,
}
`;

exports[`Resend OTP Integration Tests should resend phone OTP successfully 1`] = `
{
  "data": Any<Object>,
  "message": Any<String>,
  "statusCode": Any<Number>,
  "success": Any<Boolean>,
}
`;

exports[`Resend OTP Integration Tests should return error for invalid identifier 1`] = `
{
  "data": {},
  "message": Any<String>,
  "statusCode": 200,
  "success": true,
}
`;
